vcd/resource_vcd_nsxt_edgegateway_rate_limiting.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	_, err = nsxtEdge.UpdateQoS(qosConfig)
+	if err != nil {
+		return diag.Errorf("[rate limiting (QoS) create/update] error updating QoS configuration: %s", err)
+	}
+
+	d.SetId(edgeGatewayId)
+
+	return resourceVcdNsxtEdgegatewayRateLimitingRead(ctx, d, meta)
+}
+
+func resourceVcdNsxtEdgegatewayRateLimitingRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	orgName := d.Get("org").(string)
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+
+	nsxtEdge, err := vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+	if err != nil {
+		if govcd.ContainsNotFound(err) {
+			// When parent Edge Gateway is not found - this resource is also not found and should be
+			// removed from state
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("[rate limiting (QoS) read] error retrieving NSX-T Edge Gateway rate limiting (QoS): %s", err)
+	}
+
+	qosConfig, err := nsxtEdge.GetQoS()
+	if err != nil {
+		return diag.Errorf("[rate limiting (QoS) read] error retrieving NSX-T Edge Gateway rate limiting (QoS): %s", err)
+	}
+
+	setNsxtEdgeGatewayRateLimitingData(d, qosConfig)
+
+	return nil
+}
+
+func resourceVcdNsxtEdgegatewayRateLimitingDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	// Handling locks is conditional. There are two scenarios:
+	// * When the parent Edge Gateway is in a VDC - a lock on parent Edge Gateway must be acquired
+	// * When the parent Edge Gateway is in a VDC Group - a lock on parent VDC Group must be acquired
+	// To find out parent lock object, Edge Gateway must be looked up and its OwnerRef must be checked
+	// Note. It is not safe to do multiple locks in the same resource as it can result in a deadlock
+	parentEdgeGatewayOwnerId, _, err := getParentEdgeGatewayOwnerId(vcdClient, d)
+	if err != nil {
+		return diag.Errorf("[rate limiting (QoS) delete] error finding parent Edge Gateway: %s", err)
+	}
+
+	if govcd.OwnerIsVdcGroup(parentEdgeGatewayOwnerId) {
+		vcdClient.lockById(parentEdgeGatewayOwnerId)
+		defer vcdClient.unlockById(parentEdgeGatewayOwnerId)
+	} else {
+		vcdClient.lockParentEdgeGtw(d)
+		defer vcdClient.unLockParentEdgeGtw(d)
+	}
+
+	orgName := d.Get("org").(string)
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+
+	nsxtEdge, err := vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+	if err != nil {
+		return diag.Errorf("[rate limiting (QoS) delete] error retrieving Edge Gateway: %s", err)
+	}
+
+	// There is no real "delete" for QoS. It can only be updated to empty values (unlimited)
+	_, err = nsxtEdge.UpdateQoS(&types.NsxtEdgeGatewayQos{})
+	if err != nil {
+		return diag.Errorf("[rate limiting (QoS) delete] error updating QoS Profile: %s", err)
+	}
+
+	return nil
+}
+
+func resourceVcdNsxtEdgegatewayRateLimitingImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```

Comment:
Normally we add some docs to imports as they're usually not so trivial, not sure if it's needed here also or not?

Reply:
Added a small note